perf_bench_test.go

```diff
@@ -24,127 +24,567 @@ package main
+	}
+
+	return result
+}
+
+func dumpMetrics(t *testing.T, metrics []map[string]float64, outfile string) {
+	outFile := getOutFile(outfile)
+
+	f, err := os.OpenFile(outFile, os.O_CREATE|os.O_WRONLY, 0666)
+	require.NoError(t, err, "Failed opening file")
+	defer f.Close()
+
+	headerSet := make(map[string]bool)
+	for _, metric := range metrics {
+		for area := range metric {
+			headerSet[area] = true
+		}
 	}
 
-	log.Debugf("vmID %s: returned in %f seconds", vmIDString, time.Since(start).Seconds())
+	var headers []string
+	for area := range headerSet {
+		headers = append(headers, area)
+	}
+
+	writer := csv.NewWriter(f)
+
+	err = writer.Write(headers)
+	require.NoError(t, err, "Failed writting file")
+	writer.Flush()
+
+	for _, metric := range metrics {
+		var data []string
+		for _, header := range headers {
+			value, isPresent := metric[header]
+			if isPresent {
+				vStr := strconv.FormatFloat(value, 'f', -1, 64)
+				data = append(data, vStr)
+			} else {
+				data = append(data, "")
+			}
+		}
+		err = writer.Write(data)
+		require.NoError(t, err, "Failed writting file")
+		writer.Flush()
+	}
 }
 
-func getImages() map[string]string {
-	return map[string]string{
-		"helloworld": "vhiveease/helloworld:var_workload",
-		// "chameleon":    "vhiveease/chameleon:var_workload",
-		// "pyaes":        "vhiveease/pyaes:var_workload",
-		// "image_rotate": "vhiveease/image_rotate:var_workload",
-		// "json_serdes":  "vhiveease/json_serdes:var_workload",
-		// "lr_serving":   "vhiveease/lr_serving:var_workload",
-		// "cnn_serving":  "vhiveease/cnn_serving:var_workload",
-		// "rnn_serving":  "vhiveease/rnn_serving:var_workload",
-		// "lr_training":  "vhiveease/lr_training:var_workload",
+// getCPUIntenseRPS returns the number of requests per second that stress CPU for each image.
+func getCPUIntenseRPS() int {
+	var (
+		sum, result int
+		values      []int
+		funcs       = strings.Split(*funcNames, ",")
+		reqsPerSec  = map[string]int{
+			"helloworld":   1000,
+			"chameleon":    85,
+			"pyaes":        1000,
+			"image_rotate": 333,
+			"json_serdes":  167,
+			"lr_serving":   1000,
+			"cnn_serving":  20,
+			"rnn_serving":  100,
+		}
+	)
+
+	for _, funcName := range funcs {
+		values = append(values, reqsPerSec[funcName])
+		sum += reqsPerSec[funcName]
+	}
+
+	for _, rps := range values {
+		result += rps * rps / sum
 	}
+
+	return result
+}
+
+func getUnloadServiceTime() float64 {
+	var (
+		sum         float64
+		funcs       = strings.Split(*funcNames, ",")
+		serviceTime = map[string]float64{
+			"helloworld":   1,
+			"chameleon":    12,
+			"pyaes":        1,
+			"image_rotate": 3,
+			"json_serdes":  6,
+			"lr_serving":   1,
+			"cnn_serving":  60,
+			"rnn_serving":  10,
+		}
+	)
+
+	for _, funcName := range funcs {
+		sum += serviceTime[funcName]
+	}
+
+	return sum / float64(len(funcs))
+}
+
+func calculateRPS(vmNum int) int {
+	baseRPS := getCPUIntenseRPS()
+	return vmNum * baseRPS
+}
+
+func TestBindSocket(t *testing.T) {
```

Comment:
please move this function right after the second test/bench in this file.

Reply:
Done